### PR TITLE
[RW-7532][risk=low] Admin interface to lock/unlock a workspace

### PR DESCRIPTION
**Pull request by NehaBroad**
With this PR Admin user will be able to lock/Unlock workspace:

Unlock Workspace:

https://user-images.githubusercontent.com/34481816/144086375-aa35a12d-af0a-4594-b567-8393d8840af3.mov

Lock Workspace:

https://user-images.githubusercontent.com/34481816/144086394-581bb935-4bc7-45aa-803c-c790ec5ffdb2.mov

In case of error
Screen Shot 2021-11-30 at 11 19 12 AM

---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [x] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: ui/src/app/pages/admin/admin-workspace.tsx [hunk 1]
```diff
@@ -216,19 +217,23 @@ interface State {
...
+  loadingLockUnlockStatus: boolean;
   runtimeToDelete?: ListRuntimeResponse;
   confirmDeleteRuntime?: boolean;
   dataLoadError?: Response;
   files?: Array<FileDetail>;
+  showLockWorkspaceModal: boolean;
 }
 
-class AdminWorkspaceImpl extends React.Component<Props, State> {
+export class AdminWorkspaceImpl extends React.Component<Props, State> {
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-workspace.tsx:228]: why this change?

File: ui/src/app/pages/admin/admin-lock-request.tsx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  onClose: Function
+}
+
+export const AdminLockRequest = (props: Props) => {
+  const [requestReason, setRequestReason] = useState('');
+  const [requestDate, setRequestDate] = useState(new Date());
+  const [showError, setShowError] = useState(false);
+
+
+  const disableLockWorkspace = !requestReason ||
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-lock-request.tsx:25]: might be easier to think about this one if we define the const as `lockWorkspaceEnabled`

File: ui/src/app/pages/admin/admin-lock-request.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    const {workspace , onClose, onLock} = props;
+    const adminLockingRequest = {
+      requestReason,
+      requestDateInMillis: requestDate.valueOf()
+    }
+
+    workspaceAdminApi().setAdminLockedState(workspace, adminLockingRequest)
+      .then(() => {
+        onLock();
+        onClose();
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-lock-request.tsx:41]: not sure we want to be calling `onClose()` here.

By do so, we require the `onLock()` to be something that's appropriate to run for both locking and closing, and that may not always be true.

File: ui/src/app/pages/admin/admin-lock-request.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+         <DatePicker
+             value={requestDate}
+             placeholder='YYYY-MM-DD'
+             onChange={e => setRequestDate(e)}
+             maxDate={new Date()}
+         />
+       </div>
+     </ModalBody>
+     <ModalFooter>
+       <Button type='secondary' style={{marginRight: '0.5rem'}} onClick={() => props.onClose()}>
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-lock-request.tsx:85]: nit: onCancel?

File: ui/src/app/pages/admin/admin-lock-request.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      });
+  }
+
+   return <Modal>
+     <ModalTitle>
+       <SemiBoldHeader>Lock workspace</SemiBoldHeader>
+     </ModalTitle>
+     <ModalBody>
+       {showError && <label style={{color: colors.danger}}>
+         Something went Wrong while Locking the workspace.
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-lock-request.tsx:54]: should use lowercase w for wrong

**Review by jmthibault79 — COMMENTED** [review]
The spinner inside the button is cool!

Works well.  Just a few comments.

File: ui/src/app/pages/admin/admin-workspace.tsx [hunk 6]
```diff
@@ -216,19 +217,23 @@ interface State {
...
   constructor(props) {
     super(props);
 
     this.state = {
       workspaceDetails: {},
       cloudStorageTraffic: null,
+      loadingLockUnlockStatus: false,
```
Comment (aweng98) [inline ui/src/app/pages/admin/admin-workspace.tsx:235]: name is confusing to me: How about just `loadingAdminLockedStatus`?

File: ui/src/app/pages/admin/admin-workspace.tsx [hunk 7]
```diff
@@ -32,6 +32,7 @@ import {ReactFragment, useState} from 'react';
 import { Link, RouteComponentProps, withRouter } from 'react-router-dom';
 import {isUsingFreeTierBillingAccount} from 'app/utils/workspace-utils';
 import { EgressEventsTable } from './egress-events-table';
+import {AdminLockRequest} from './admin-lock-request';
```
Comment (aweng98) [inline ui/src/app/pages/admin/admin-workspace.tsx:35]: nit: relative path in import. can we use the absolute path `app/pages/admin/admin-lock-request`?

File: api/src/main/java/org/pmiops/workbench/api/WorkspaceAdminController.java [hunk 8]
```diff
@@ -98,13 +98,13 @@ public WorkspaceAdminController(WorkspaceAdminService workspaceAdminService) {
           String.format("Cannot have empty Request reason or Request Date"));
     }
     workspaceAdminService.setAdminLockedState(workspaceNamespace, lockingRequest);
-    return ResponseEntity.ok().build();
+    return ResponseEntity.ok(new EmptyResponse());
```
Comment (aweng98) [inline api/src/main/java/org/pmiops/workbench/api/WorkspaceAdminController.java:101]: 👍🏻

File: ui/src/app/pages/admin/admin-workspace.tsx (hunk 7 above)
Comment (jmthibault79) [reply ui/src/app/pages/admin/admin-workspace.tsx:35]: I dunno.  This is a very common pattern, especially when it's a related component like it is here.

File: ui/src/app/pages/admin/admin-workspace.tsx (hunk 1 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-workspace.tsx:228]: I was writing test for this class but realized its going to take some time. I added export  just for that

File: ui/src/app/pages/admin/admin-lock-request.tsx (hunk 3 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-lock-request.tsx:41]: Good point

File: ui/src/app/pages/admin/admin-lock-request.tsx (hunk 2 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-lock-request.tsx:25]: For me it was easier to do it the other way around :) i can give it a try

File: ui/src/app/pages/admin/admin-workspace.tsx (hunk 6 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-workspace.tsx]: Done

File: ui/src/app/pages/admin/admin-lock-request.tsx (hunk 5 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-lock-request.tsx]: Done

File: ui/src/app/pages/admin/admin-lock-request.tsx (hunk 4 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-lock-request.tsx]: Done

File: ui/src/app/pages/admin/admin-lock-request.tsx (hunk 2 above)
Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-lock-request.tsx]: Done also added a condition to confirm the date is valid or not

Comment (jmthibault79) [reply ui/src/app/pages/admin/admin-lock-request.tsx]: nice, thanks

File: ui/src/app/pages/admin/admin-lock-request.tsx [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  const [requestReason, setRequestReason] = useState('');
+  const [requestDate, setRequestDate] = useState(new Date());
+  const [showError, setShowError] = useState(false);
+
+  const enableLockButton = requestReason?.length > 0 && requestDate?.toString() !== ''
+      && !isNaN(requestDate.valueOf()) && !showError;
+
+
+  const getToolTipContent = showError ? 'Error occurred while Locking Workspace' :
+      'Request Reason & Valid Request Date (in YYYY-MM-DD Format) are required to lock workspace';
```
Comment (jmthibault79) [inline ui/src/app/pages/admin/admin-lock-request.tsx:29]: the user does not control whether the date is valid because they are picking from a date widget

Comment (NehaBroad) [reply ui/src/app/pages/admin/admin-lock-request.tsx:29]: The date input box is editable, i was able to enter dates like 2021-13-01 or just 2021-13

Comment (jmthibault79) [reply ui/src/app/pages/admin/admin-lock-request.tsx:29]: sorry, you're right

**Review by jmthibault79 — APPROVED** [review]
